### PR TITLE
Deepseek v2 16b enablement

### PR DESCRIPTION
**Pull request by hakankiymaz-amd**
Deepseek v2 16b enablement
[test_report.csv](https://github.com/user-attachments/files/18084277/test_report.csv)
[MI300X-DeepSeek-V2-Lite-bf16-seq2048-tp1pp1ep8-mbsgbs-ac_sel-do_true-fa_true-sp_true-20241210_173933.log](https://github.com/user-attachments/files/18084278/MI300X-DeepSeek-V2-Lite-bf16-seq2048-tp1pp1ep8-mbsgbs-ac_sel-do_true-fa_true-sp_true-20241210_173933.log)

throughput per GPU: 611.822
All tests passed

### CODE REVIEW

File: megatron/core/transformer/transformer_layer.py [hunk 1]
```diff
@@ -281,6 +282,7 @@ def forward(
         attention_output_with_bias = self.self_attention(
             input_layernorm_output,
             attention_mask=attention_mask,
+            #position_ids=position_ids,
```
Comment (lcskrishna) [inline megatron/core/transformer/transformer_layer.py]: remove this if its not needed.

File: megatron/training/utils.py [hunk 2]
```diff
@@ -228,6 +229,210 @@ def get_ltor_masks_and_position_ids(data,
 
     return attention_mask, loss_mask, position_ids
 
+def get_batch_on_this_tp_rank_original(data_iterator):
```
Comment (wenchenvincent) [inline megatron/training/utils.py]: What's the difference between this function and the `get_batch_on_this_tp_rank` function in this file?

File: megatron/core/models/deepseekv2/yarn_rotary_pos_embedding.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+def yarn_linear_ramp_mask(min, max, dim):
+    if min == max:
+        max += 0.001  # Prevent singularity
+
+    linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
+    ramp_func = torch.clamp(linear_func, 0, 1)
+    return ramp_func
+
+class DeepseekV2YarnRotaryEmbedding(DeepseekV2RotaryEmbedding):
```
Comment (wenchenvincent) [inline megatron/core/models/deepseekv2/yarn_rotary_pos_embedding.py]: Can we use `YarnRotaryEmbedding` from `megatron.core.models.common.embeddings.yarn_rotary_pos_embedding` instead?

File: megatron/core/models/deepseekv2/transformer_layer.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        # Residual connection.
+        residual = hidden_states
+
+        # Optional Input Layer norm
+        input_layernorm_output = self.input_layernorm(hidden_states)
+
+        # Self attention.
+        attention_output_with_bias = self.self_attention(
+            input_layernorm_output,
+            position_ids=position_ids,
```
Comment (wenchenvincent) [inline megatron/core/models/deepseekv2/transformer_layer.py]: It seems the main difference between the `forward` functions of `DeekSeekv2TransformerLayer` and `TransformerLayer` is the extra argument of `position_ids` in this function call. I am wondering if we can refactor it to reduce the code duplication.

File: megatron/core/models/deepseekv2/transformer_block.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
+import re
```
Comment (wenchenvincent) [inline megatron/core/models/deepseekv2/transformer_block.py]: Can we use megatron/core/models/transformer_block/transformer_block.py instead?

File: megatron/training/utils.py (hunk 2 above)
Comment (hakankiymaz-amd) [reply megatron/training/utils.py]: fixed

File: examples/deepseek_v2/pretrain_deepseek.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
+import os
```
Comment (wenchenvincent) [inline examples/deepseek_v2/pretrain_deepseek.py:8]: Could you add the copyright statement to this file?

File: megatron/core/models/deepseekv2/model.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
+import logging
```
Comment (wenchenvincent) [inline megatron/core/models/deepseekv2/model.py:3]: Could you add the copyright statement?

File: megatron/training/utils.py [hunk 8]
```diff
@@ -229,6 +230,111 @@ def get_ltor_masks_and_position_ids(data,
...
 
 
+def get_batch_on_this_tp_rank_idxmap_sft(data_iterator):
+    args = get_args()
+    tokenizer = get_tokenizer()
+    def _broadcast(item):
+        if item is None:
+            return
+        torch.distributed.broadcast(item, mpu.get_tensor_model_parallel_src_rank(),
+                                    group=mpu.get_tensor_model_parallel_group())
```
Comment (wenchenvincent) [inline megatron/training/utils.py:253]: Nit: Make the formatting a bit prettier.

**Comment by wenchenvincent** [comment]
@hakankiymaz-amd Could you also make sure the copyright statements are all applied properly? If files are copied over from PAI-megatron-patch without any change, you should keep the original copyright statement there. If it is incorporated with your change, You need to add additional copyright statement for AMD and specify what are the changes.

**Comment by hakankiymaz-amd** [comment]
Copyright statements are added proper to changes from PAI and changes from AMD.

**Review by lcskrishna — APPROVED** [review]
LGTM, @wenchenvincent Could you please re-review again to see if there is something that we might have missed.

File: megatron/legacy/model/rms_norm.py [hunk 9]
```diff
@@ -6,21 +6,21 @@
 class RMSNorm(torch.nn.Module):
 
     def __init__(self,
-                 dim: int,
+                 hidden_size: int,
```
Comment (wenchenvincent) [inline megatron/legacy/model/rms_norm.py:9]: Why do we need to change the argument name here?

Comment (hakankiymaz-amd) [reply megatron/legacy/model/rms_norm.py:9]: I have got a few errors with **dim** as argument name. 
In layer_specs.py  MLASelfAttention , there are [q_layernorm](https://github.com/ROCm/Megatron-LM/blob/bb93ccbfeae6363c67b361a97a27c74ab86e7e92/megatron/core/transformer/multi_latent_attention.py#L265), [kv_layernorm](https://github.com/ROCm/Megatron-LM/blob/bb93ccbfeae6363c67b361a97a27c74ab86e7e92/megatron/core/transformer/multi_latent_attention.py#L272).   
These are built using hidden_size. That is why I have changed.

File: megatron/core/models/deepseekv2/layer_specs.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    )
+    mlp_dense = _get_mlp_module_spec(
+        use_te=False, num_experts=None, moe_grouped_gemm=moe_grouped_gemm
+    )
+
+    return ModuleSpec(
+        module=TransformerLayer,
+        submodules=TransformerLayerSubmodules(
+            self_attention=ModuleSpec(
+                module=DeepSeekv2SelfAttention,
```
Comment (wenchenvincent) [inline megatron/core/models/deepseekv2/layer_specs.py]: It seems that for the implementation with Transformer Engine modules, we use `DeepSeekv2SelfAttention` with `TEDotProductAttentionMLA` while for the implementation with Megatron modules, we use `MLASelfAttention` with `TEDotProductAttention`. I am wondering if we can also use `MLASelfAttention` with `TEDotProductAttention` for the implementation with Transformer Engine modules.

**Comment by billishyahao** [comment]
Hi @hakankiymaz-amd , Thank you for the Deepseek v2 enablement effort. Do we have any comparison data with H100 internally?

**Comment by lcskrishna** [comment]
> Hi @hakankiymaz-amd , Thank you for the Deepseek v2 enablement effort. Do we have any comparison data with H100 internally?

@billishyahao Could you please reach out internally.

**Comment by wenchenvincent** [comment]
Closing this PR as another newer PR for the same purpose was merged.
